skorch/probabilistic.py

```diff
@@ -0,0 +1,941 @@
+
+        """
+        self.check_is_fitted()
+        samples = [p.sample(torch.Size([n_samples])) for p in self.forward_iter(X)]
+        return torch.cat(samples, axis=axis)
+
+    def confidence_region(self, X, sigmas=2):
+        """Returns 2 standard deviations above and below the mean.
+
+        X : input data
+          The samples where the GP is evaluated.
+
+        sigmas : int (default=2)
+          The number of standard deviations of the region.
+
+        Returns
+        -------
+        lower : torch.Tensor
+          The lower end of the confidence region.
+
+        upper : torch.Tensor
+          The upper end of the confidence region.
+
+        """
+        nonlin = self._get_predict_nonlinearity()
+        lower, upper = [], []
+        for yi in self.forward_iter(X):
+            posterior = yi[0] if isinstance(yi, tuple) else yi
+            mean = posterior.mean
+            std = posterior.stddev
+            std = std.mul_(sigmas)
+            lower.append(nonlin(mean.sub(std)))
+            upper.append(nonlin(mean.add(std)))
+
+        lower = torch.cat(lower)
+        upper = torch.cat(upper)
+        return lower, upper
+
+    def __getstate__(self):
+        try:
+            return super().__getstate__()
+        except pickle.PicklingError as exc:
+            msg = ("This GPyTorch model cannot be pickled. The reason is probably this:"
+                   " https://github.com/pytorch/pytorch/issues/38137. "
+                   "Try using 'dill' instead of 'pickle'.")
+            raise pickle.PicklingError(msg) from exc
+
+
+class _GPRegressorPredictMixin:
+    """Mixin class that provides a predict method for GP regressors."""
+    def predict(self, X, return_std=False, return_cov=False):
+        """Returns the predicted mean and optionally standard deviation.
+
+        Parameters
+        ----------
+        X : input data
+          Input data where the GP is evaluated.
+
+        return_std : bool (default=False)
+          If True, the standard-deviation of the predictive distribution at the
+          query points is returned along with the mean.
+
+        return_cov : bool (default=False)
+          This exists solely for sklearn compatibility and is not supported by
+          skorch.
+
+        Returns
+        -------
+        y_pred : numpy ndarray
+          Mean of predictive distribution at the query points.
+
+        y_std : numpy ndarray
+          Standard deviation of predictive distribution at query points. Only
+          returned when ``return_std`` is True.
+
+        """
+        if return_cov:
+            msg = ("The 'return_cov' argument is not supported. Please try: "
+                   "'posterior = next(gpr.forward_iter(X)); "
+                   "posterior.covariance_matrix'.")
+            raise NotImplementedError(msg)
+
+        nonlin = self._get_predict_nonlinearity()
+        y_preds, y_stds = [], []
+        for yi in self.forward_iter(X, training=False):
```

Comment:
One thing that GPyTorch does is we cache a couple things that don't depend on the test points the first time we make predictions so that the next time we make predictions we don't have O(n^3) complexity. These caches get deleted in some cases, like if the model gets put back in `training` mode -- I assume that `forward_iter` wouldn't do any of these things though.

It should be pretty easy to test:
```python
gpr = ...
gpr.fit(...)
with gpytorch.settings.fast_pred_var():
    y_pred, y_std = gpr.predict(test_x, return_std=True)
    # test_x could change here.
    y_pred, y_std = gpr.predict(test_x, return_std=True)   # This call should be significantly faster.

Reply:
Indeed, when using `forward_iter`, there is no risk of the model being set to `training` again. Therefore, I'll add the `fast_pred_var` context for it. When testing, I didn't find any noticeable speed difference. Perhaps, that is because I'm mostly using toy models on small data in the notebook?